include: "bqsr_bwa_align.smk"
include: "make_directories.smk"

rule filter:
    input:
        "{out}/{vcf}/combined_recal.vcf".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"])
    output:
        "{out}/{vcf}/combined_filt.vcf".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"])
    shell:
        r"grep '^#\|PASS' {input} > {output}"

rule indel_apply:
    input:
        "{out}/{vcf}/combined_snp_recal.vcf".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"]),
        "{out}/{vcf}/combined_indels.recal".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"]),
        "{out}/{vcf}/combined_indels.tranches".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"]),
        gatk=config['GATK'],
        ref=config['Reference']
    output:
        "{out}/{vcf}/combined_recal.vcf".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"])
    threads:
        4
    shell:
        "java -Xmx20g -Djava.io.tmpdir=/mnt/data/jeremy/tmp -jar {input.gatk} -T ApplyRecalibration -nt {threads} -R {input.ref} -input {input[0]} --mode INDEL -ts_filter_level 98.0 "
        "-recalFile {input[1]} -tranchesFile {input[2]} -o {output}"

rule indel_recal:
    input:
        "{out}/{vcf}/combined_snp_recal.vcf".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"]),
        gatk=config['GATK'],
        ref=config['Reference'],
        dbsnp=config['resource']['dbsnp'],
        mills=config['resource']['mills']
    output:
        "{out}/{vcf}/combined_indels.recal".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"]),
        "{out}/{vcf}/combined_indels.tranches".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"]),
    threads:
        6
    shell:
        "java -Xmx20g -Djava.io.tmpdir=/mnt/data/jeremy/tmp -jar {input.gatk} -T VariantRecalibrator -nt {threads} -R {input.ref} -input {input[0]} "
        "-resource:mills,known=true,training=true,truth=true,prior=12.0 {input.mills} "
        "-resource:dbsnp,known=true,training=false,truth=false,prior=2.0 {input.dbsnp} "
        "--mode INDEL --recal_file {output[0]} --tranches_file {output[1]} -an DP -an QD -an FS -an ReadPosRankSum -an SOR -an MQRankSum "
        "-an MQ -tranche 100.0 -tranche 99.99 -tranche 99.98 -tranche 99.97 -tranche 99.96 -tranche 99.95 -tranche 99.93 -tranche 99.90 "
        "-tranche 99.8 -tranche 99.5 -tranche 99 -tranche 98 -tranche 97.5 -tranche 97 -tranche 96 -tranche 95 -tranche 90 -mG 5 "

rule snp_apply:
    input:
        "{out}/{vcf}/combined_raw.vcf".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"]),
        "{out}/{vcf}/combined_snps.recal".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"]),
        "{out}/{vcf}/combined_snps.tranches".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"]),
        gatk=config['GATK'],
        ref=config['Reference']
    output:
        #"Data_Files/VCF/fam_snp_recal.vcf"
        "{out}/{vcf}/combined_snp_recal.vcf".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"])
    threads:
        4
    shell:
        "java -Xmx20g -Djava.io.tmpdir=/mnt/data/jeremy/tmp -jar {input.gatk} -T ApplyRecalibration -nt {threads} -R {input.ref} "
        "-input {input[0]} --mode SNP --ts_filter_level 99.0 -recalFile {input[1]} "
        "-tranchesFile {input[2]} -o {output}"

rule snp_recal:
    input:
        "{out}/{vcf}/combined_raw.vcf".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"]),
        gatk=config['GATK'],
        ref=config['Reference'],
        dbsnp=config['resource']['dbsnp'],
        hapmap=config['resource']['hapmap'],
        omni=config['resource']['omni'],
        onekg=config['resource']['onekgenomes']
    output:
        "{out}/{vcf}/combined_snps.recal".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"]),
        "{out}/{vcf}/combined_snps.tranches".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"])
    threads:
        6
    shell:
        "java -Xmx20g -Djava.io.tmpdir=/mnt/data/jeremy/tmp -jar {input.gatk} -T VariantRecalibrator -nt {threads} -R {input.ref} "
        "-input {input[0]} -resource:hapmap,known=false,training=true,truth=true,prior=15.0 {input.hapmap} "
        "-resource:omni,known=false,training=true,truth=true,prior=12.0 {input.omni} "
        "-resource:1000G,known=false,training=true,truth=false,prior=10.0 {input.onekg} "
        "-resource:dbsnp,known=true,training=false,truth=false,prior=2.0 {input.dbsnp} "
        "--mode SNP --recal_file {output[0]} --tranches_file {output[1]} "
        "-an DP -an QD -an FS -an ReadPosRankSum -an SOR -an MQRankSum -an MQ -tranche 100.0 -tranche 99.99 -tranche 99.98 -tranche 99.97 -tranche 99.96 "
        "-tranche 99.95 -tranche 99.93 -tranche 99.90 -tranche 99.8 -tranche 99.5 -tranche 99 -tranche 98 -tranche 90 -mG 5"

rule bgzip:
    input:
        vcf="{out}/{tmp}/combined_raw.vcf".format(out=config["dir"]["out"], tmp=config["dir"]["tmp"])
    output:
        vcfgz="{out}/{vcf}/combined_raw.vcf.gz".format(out=config["dir"]["out"], vcf=config["dir"]["vcf"])
    threads:
        1
    run:
        shell("bgzip -c {input.vcf} > {output.vcfgz}")

rule genotype:
    input:
        gvcfs=("{out}/{gvcf}/{sample}.g.vcf".format(out=config["dir"]["out"], gvcf=config["dir"]["gvcf"], sample=sample) for sample in config["Samples"].keys()),
        gatk=config['GATK'],
        ref=config['Reference'],
        dbsnp=config['resource']['dbsnp']
    output:
        "{out}/{tmp}/combined_raw.vcf".format(out=config["dir"]["out"], tmp=config["dir"]["tmp"])
    threads:
        1
    run:
        gvcfs = '-V ' + ' -V '.join(input.gvcfs)
        shell("java -Xmx32g -jar {input.gatk} -T GenotypeGVCFs -nt {threads} "
        "-R {input.ref} {gvcfs} -D {input.dbsnp} "
        "-o {output} "
        "-G StandardAnnotation -A AlleleBalance -A TandemRepeatAnnotator "
        "-A ClippingRankSumTest -A GCContent -A MappingQualityZero -A SpanningDeletions -A StrandOddsRatio; sleep 2")

rule call_vars:
    input:
        "{out}/{bam}/{sample}_recal.bam".format(out=config["dir"]["out"], bam=config["dir"]["bam"], sample="{sample}"),
        "{out}/{bam}/{sample}_recal.bai".format(out=config["dir"]["out"], bam=config["dir"]["bam"], sample="{sample}"),
        gatk=config['GATK'],
        ref=config['Reference'],
    output:
        "{out}/{gvcf}/{sample}.g.vcf".format(out=config["dir"]["out"], gvcf=config["dir"]["gvcf"], sample="{sample}")
    params:
        exome_region = config['exome_region']
    threads:
        1
    run:
        outfile = "{out}/{gvcf}/{sample}.g.vcf".format(out=config["dir"]["out"], gvcf=config["dir"]["gvcf"], sample="{wildcards.sample}")
        if str(config['exome_only'])=='true':
            exome_region = "-L " + params.exome_region
        else:
            exome_region = ""
        print(config['exome_only'])
        shell(
                "java -Xmx4g -jar {input.gatk} -T HaplotypeCaller -R {input.ref} "
                "-I {input[0]} -o {output} "
                "{exome_region} " 
                "--emitRefConfidence GVCF "
                "--variant_index_type LINEAR --variant_index_parameter 128000 "
                "-G StandardAnnotation -A AlleleBalance -A TandemRepeatAnnotator "
                "-A ClippingRankSumTest -A GCContent "
                "-A MappingQualityZero -A SpanningDeletions "
                "-A StrandOddsRatio; sleep 2"
            )

